### PR TITLE
[DeviceSanitizer] Support check invalid kernel argument

### PR DESCRIPTION
**Pull request by AllanZyne**
LLVM: https://github.com/intel/llvm/pull/14512

### CODE REVIEW

**Review by yingcong-wu — APPROVED** [review]
lgtm

File: source/loader/layers/sanitizer/asan_validator.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    if (AllocInfo->Context != Context) {
+        return ValidateUSMResult::fail(ValidateUSMResult::BAD_CONTEXT,
+                                       AllocInfo);
+    }
+
+    if (!IsSameDevice(AllocInfo->Device, Device)) {
+        return ValidateUSMResult::fail(ValidateUSMResult::BAD_DEVICE,
+                                       AllocInfo);
+    }
```
Comment (AllanZyne) [inline source/loader/layers/sanitizer/asan_validator.cpp:62]: Note: host usm support on every devices

Comment (yingcong-wu) [reply source/loader/layers/sanitizer/asan_validator.cpp:62]: I think FIXME notes should write into source code file rather than PR page.

**Comment by AllanZyne** [comment]
force push to reset the code to merge latest syclos tag

**Comment by AllanZyne** [comment]
Hi @oneapi-src/unified-runtime-maintain, this PR is needed in 2025.0, please review! thanks!

**Review by pbalcer — APPROVED** [review]

File: source/loader/layers/sanitizer/asan_report.cpp [hunk 2]
```diff
@@ -157,4 +154,47 @@ void ReportUseAfterFree(const DeviceSanitizerReport &Report,
...
+                                 uptr Addr, const ValidateUSMResult &VR,
+                                 StackTrace Stack) {
+    getContext()->logger.always("\n====ERROR: DeviceSanitizer: "
+                                "invalid-argument on kernel <{}>",
+                                DemangleName(GetKernelName(Kernel)));
+    Stack.print();
+    auto &AI = VR.AI;
+    switch (VR.Type) {
+    case ValidateUSMResult::MAYBE_HOST_POINTER:
+        getContext()->logger.always("The {}th argument {} is not a USM pointer",
```
[the remark this answers is not in the document]
Comment (AllanZyne) [reply source/loader/layers/sanitizer/asan_report.cpp:167]: Thanks for reminding me :)
I fall into programmer's mindset.

**Comment by AllanZyne** [comment]
CI failures are unrelated

**Pull request merged**